src/Core/Animation/KeyFramedValue.hpp

```diff
@@ -0,0 +1,215 @@
+#ifndef RADIUMENGINE_KEYFRAMEDVALUE_HPP
+#define RADIUMENGINE_KEYFRAMEDVALUE_HPP
+
+#include <map>
+#include <set>
+
+#include <Core/Containers/AlignedAllocator.hpp>
+#include <Core/Math/Math.hpp>
+#include <Core/Types.hpp>
+
+namespace Ra {
+namespace Core {
+namespace Animation {
+
+/**
+ * KeyFramedValueBase defines the following concept:
+ * any variable (e.g. transformation matrix, color, scalar value, flag) that
+ * needs to be animated can be controlled by a `KeyFramedValue`, which inherits
+ * KeyFramedValueBase (see dedicated documentation).
+ *
+ * KeyFramedValueBase defines the basic interface available to control keyframes:
+ * insertKeyFrame, removeKeyFrame and moveKeyFrame.
+ */
+class RA_CORE_API KeyFramedValueBase
+{
+  public:
+    KeyFramedValueBase() {}
+
+    virtual ~KeyFramedValueBase() {}
+
+    /**
+     * \returns the number of keyframes.
+     */
+    virtual size_t size() const = 0;
+
+    /**
+     * \returns the ordered list of the points in time where a keyframe is defined.
+     */
+    virtual inline std::vector<Scalar> getTimes() const = 0;
+
+    /**
+     * Inserts a keyframe at time \p t.
+     */
+    virtual void insertKeyFrame( const Scalar& t ) = 0;
+
+    /**
+     * Removes the \p i-th keyframe.
+     * \returns true if the keyframe has been removed, false otherwise.
+     */
+    virtual bool removeKeyFrame( size_t i ) = 0;
```

Comment:
There must be at least one keyframe so that the KeyFramedValue is well defined, will enhance the doc

Reply:
Sure you need at least some values for the interpolator to function. But the number will depend on the interpolator (1 for nearest, 2 for linear, 3 for cubic, ...). So, this is related on the way the interpolator uses the values, not to the values themselves. And an empty keyframed value is valid as long as we consider it is a collection of <time-value> pair even if it can't be used because it is empty.
So, the only case of failure will be on an empty collection, and this is better manage by a precondition/assertion.